components/Palettes/src/UIColor+MDCPalettable.m

```diff
@@ -0,0 +1,165 @@
+/*
+ Copyright 2017-present the Material Components for iOS authors. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+#import <CoreGraphics/CoreGraphics.h>
+#import <UIKit/UIKit.h>
+#import <objc/runtime.h>
+
+#import "MDCPalettes.h"
+#import "UIColor+MDCPalettable.h"
+
+@implementation UIColor (MDCPalettable)
+
+static BOOL ColorEqualToColor(UIColor *color, UIColor *otherColor) {
+  if ((color == nil && otherColor != nil) ||
+      (color != nil && otherColor == nil)) {
+    return NO;
+  }
+  if (!color && !otherColor) {
+    return YES;
+  }
+  size_t numComponents = CGColorGetNumberOfComponents(color.CGColor);
+  size_t otherNumComponents = CGColorGetNumberOfComponents(otherColor.CGColor);
+  if (numComponents != otherNumComponents) {
+    return NO;
+  }
+
+  const CGFloat *components = CGColorGetComponents(color.CGColor);
+  const CGFloat *otherComponents = CGColorGetComponents(otherColor.CGColor);
+  for (size_t i = 0; i < numComponents + 1; ++i) {
+    if (fabs(components[i] - otherComponents[i]) > 0.001) {
+      return NO;
+    }
+  }
+  return YES;
+}
+
++ (MDCPalette *)paletteForColor:(UIColor *)color
+                    darkerColor:(UIColor * __autoreleasing*)darkerColor
```

Comment:
Why you need `__autoreleasing` here? Actually asking. I never use it.

Reply:
**Apprentice**

This [StackOverflow thread talks about why to put the `__autoreleasing` attribute on the method parameters](https://stackoverflow.com/questions/23268576/autoreleasing-in-errornserror-autoreleasing-outerror).  Basically, it's making explicit what ARC will do automatically (for this non-templated parameter).  Since Apple usually makes it explicit for its in-out parameters, I like to do so as well.

**Journeyperson**
A second [StackOverflow thread has a pretty good answer about what happens when a local variable is passed-by-reference to a function as an in-out parameter](https://stackoverflow.com/questions/8862023/in-which-situations-do-we-need-to-write-the-autoreleasing-ownership-qualifier).  

**Master**
This third [StackOverflow thread has a very detailed discussion of what happens to all the objects](https://stackoverflow.com/questions/8814718/handling-pointer-to-pointer-ownership-issues-in-arc).

Reply:
TY!